src/haz3lcore/statics/Statics.re

```diff
@@ -522,141 +331,142 @@ and upat_to_info_map =
   | String(_) => atomic(Just(String))
   | ListLit([]) => atomic(Just(List(Unknown(Internal))))
   | ListLit(ps) =>
-    let modes = List.init(List.length(ps), _ => Typ.matched_list_mode(mode));
-    let p_ids = List.map(Term.UPat.rep_id, ps);
-    let (ctx, infos) =
-      List.fold_left2(
-        ((ctx, infos), e, mode) => {
-          let (_, ctx, _) as info = upat_to_info_map(~ctx, ~mode, e);
-          (ctx, infos @ [info]);
-        },
-        (ctx, []),
-        ps,
-        modes,
-      );
-    let tys = List.map(((ty, _, _)) => ty, infos);
-    let self: Typ.self =
-      switch (Typ.join_all(tys)) {
-      | None =>
-        Joined(
-          ty => List(ty),
-          List.map2((id, ty) => Typ.{id, ty}, p_ids, tys),
-        )
-      | Some(ty) => Just(List(ty))
-      };
-    let info: t = InfoPat({cls, self, mode, ctx, term: upat});
-    let m = union_m(List.map(((_, _, m)) => m, infos));
-    /* Add an entry for the id of each comma tile */
-    let m = List.fold_left((m, id) => Id.Map.add(id, info, m), m, ids);
-    (typ_after_fix(mode, self), ctx, m);
+    let modes = Typ.matched_list_lit_modes(mode, List.length(ps));
+    let (ctx, tys, m) = ctx_fold(ctx, m, ps, modes);
+    add(
+      ~self=Info.join(ty => List(ty), tys, List.map(UPat.rep_id, ps), ctx),
+      ~ctx,
+      m,
+    );
   | Cons(hd, tl) =>
-    let mode_e = Typ.matched_list_mode(mode);
-    let (ty1, ctx, m_hd) = upat_to_info_map(~ctx, ~mode=mode_e, hd);
-    let (_, ctx, m_tl) = upat_to_info_map(~ctx, ~mode=Ana(List(ty1)), tl);
-    add(~self=Just(List(ty1)), ~ctx, union_m([m_hd, m_tl]));
-  | Tag(name) =>
-    switch (BuiltinADTs.get_tag_typ(name)) {
-    | None => atomic(Free(Tag))
-    | Some(typ) => atomic(Just(typ))
-    }
+    let (hd, m) = go(~ctx, ~mode=Typ.matched_list_mode(mode), hd, m);
+    let (tl, m) = go(~ctx=hd.ctx, ~mode=Ana(List(hd.ty)), tl, m);
+    add(~self=Just(List(hd.ty)), ~ctx=tl.ctx, m);
   | Wild => atomic(Just(unknown))
   | Var(name) =>
-    let typ = typ_after_fix(mode, Just(Unknown(Internal)));
-    let entry = Ctx.VarEntry({name, id: Term.UPat.rep_id(upat), typ});
-    add(~self=Just(unknown), ~ctx=Ctx.extend(entry, ctx), Id.Map.empty);
+    /* Note the self type assigned to pattern variables (unknown)
```

Comment:
not sure I understand this -- shouldn't the expected type be added to the context when the mode is analytic?

Reply:
That is what's happening, it's just slightly indirect. ty_after_fix takes the mode and handles this logic; it also handles the Syn/SynFun case, and the logic is generic.